agb/src/agb_alloc/block_allocator.rs

```diff
@@ -78,26 +132,52 @@ unsafe impl GlobalAlloc for BlockAllocator {
                         .add(block_after_layout_offset)
                         .cast();
                     *split_ptr = split_block;
-                    *list_ptr = NonNull::new(split_ptr);
+                    *list_ptr = NonNull::new(split_ptr).map(SendNonNull);
 
                     return curr.as_ptr().cast();
                 }
                 current_block = curr_block.next;
                 list_ptr = &mut curr_block.next;
             }
-        }
 
-        self.new_block(layout)
+            self.new_block(layout, key)
+        })
     }
 
     unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
-        let new_layout = Block::either_layout(layout);
-        let mut state = self.state.lock();
-        let new_block_content = Block {
-            size: new_layout.size(),
-            next: state.first_free_block,
-        };
-        *ptr.cast() = new_block_content;
-        state.first_free_block = NonNull::new(ptr.cast());
+        let new_layout = Block::either_layout(layout).pad_to_align();
+        free(|key| {
+            let mut state = self.state.borrow(*key).borrow_mut();
+
+            let mut list_ptr = &mut state.first_free_block;
+
+            loop {
+                match list_ptr {
+                    Some(mut current_block) => {
+                        if current_block.as_ptr().cast() > ptr {
+                            let new_block_content = Block {
+                                size: new_layout.size(),
+                                next: Some(current_block),
+                            };
+                            *ptr.cast() = new_block_content;
+                            *list_ptr = NonNull::new(ptr.cast()).map(SendNonNull);
```

Comment:
I don't understand what this line is doing (or the same one in the None case)...